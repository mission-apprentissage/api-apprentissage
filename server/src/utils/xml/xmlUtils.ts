import { createStream, QualifiedAttribute } from "sax";
import { Readable, Transform } from "stream";

interface XmlJson {
  attributes: Record<string, string | QualifiedAttribute>;
  children: Record<string, XmlJson[]>;
  value?: string;
}

// Parses an XML file stream and transform it into a object stream with the given selector
// This function is useful to parse large XML files without loading the entire file into memory
export function parseXmlFileStream(input: Readable, selector: string): Readable {
  const saxStream = createStream(true, { trim: true });
  const outputStream = new Transform({ objectMode: true });

  const stack: XmlJson[] = [];

  const safeListener = <T>(listener: (v: T) => unknown) => {
    return (v: T): void => {
      try {
        listener(v);
      } catch (error) {
        outputStream.emit("error", error);
      }
    };
  };

  input.on("error", (error) => {
    outputStream.emit("error", error);
    outputStream.destroy();
  });

  saxStream.on("error", (error) => {
    outputStream.emit("error", error);
    outputStream.destroy();
  });

  saxStream.on(
    "opentag",
    safeListener(({ name, attributes }) => {
      const currentTag: XmlJson = { attributes, children: {} };

      if (stack.length === 0) {
        if (name !== selector) {
          return;
        }

        stack.push(currentTag);
        return;
      }

      const parentTag = stack[stack.length - 1];
      if (parentTag.children[name]) {
        parentTag.children[name].push(currentTag);
      } else {
        parentTag.children[name] = [currentTag];
      }
      stack.push(currentTag);
    })
  );
  saxStream.on(
    "closetag",
    safeListener(() => {
      const tag = stack.pop();
      if (tag && stack.length === 0) {
        outputStream.push(tag);
      }
    })
  );
  saxStream.on(
    "text",
    safeListener((tag) => {
      if (stack.length > 0) {
        stack[stack.length - 1].value = tag;
      }
    })
  );
  saxStream.on("end", () => {
    outputStream.push(null);
  });

  input.pipe(saxStream);

  return outputStream;
}
