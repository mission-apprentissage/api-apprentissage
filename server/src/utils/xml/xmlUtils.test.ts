import { createReadStream } from "fs";
import { dirname, join } from "path";
import { Readable } from "stream";
import { fileURLToPath } from "url";
import { describe, expect, it } from "vitest";

import { parseXmlFileStream } from "./xmlUtils.js";

describe("parseXmlFileStream", () => {
  it("should parse XML file stream and emit open tags", async () => {
    const dataFixture = join(dirname(fileURLToPath(import.meta.url)), `fixtures/file.xml`);

    const xmlFileStream = createReadStream(dataFixture);
    const outputStream = parseXmlFileStream(xmlFileStream, "FICHE");

    const data = [];

    for await (const chunk of outputStream) {
      data.push(chunk);
    }

    expect(data).toMatchSnapshot();
  });

  it("should handle errors during XML parsing", async () => {
    const expectedError = new Error("Error reading XML file");
    function* generateXmlStreamWithReadError() {
      yield '<?xml version="1.0" encoding="UTF-8" standalone="no"?>';
      yield "<FICHES><VERSION_FLUX>3.0</VERSI";
      yield "ON_FLUX><FICHE>";
      yield '<ID_FICHE yop="123">22192</ID_FICHE>';

      throw expectedError;
    }
    const xmlFileStream = Readable.from(generateXmlStreamWithReadError());

    const outputStream = parseXmlFileStream(xmlFileStream, "FICHE");

    const error = await new Promise((resolve) => {
      outputStream.on("error", resolve);
    });

    expect(error).toBe(expectedError);
  });

  it("should handle invalid XML file", async () => {
    function* generateXmlStreamWithReadError() {
      yield '<?xml version="1.0" encoding="UTF-8" standalone="no"?>';
      yield "<FICHES><VERSION_FLUX>3.0</VERSI";
      yield "ON_FLUX><FICHE>";
      yield '<ID_FICHE yop="123">22192</ID_FICHE>';
    }
    const xmlFileStream = Readable.from(generateXmlStreamWithReadError());

    const outputStream = parseXmlFileStream(xmlFileStream, "FICHE");

    const error = await new Promise((resolve) => {
      outputStream.on("error", resolve);
    });

    expect((error as Error).message).toMatch("Unclosed root tag");
  });
});
