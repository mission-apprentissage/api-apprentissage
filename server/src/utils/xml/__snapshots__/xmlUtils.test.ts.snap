// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseXmlFileStream > should parse XML file stream and emit open tags 1`] = `
[
  {
    "attributes": {},
    "children": {
      "ACTIF": [
        {
          "attributes": {},
          "children": {},
          "value": "Non",
        },
      ],
      "ETAT_FICHE": [
        {
          "attributes": {},
          "children": {},
          "value": "Publiée",
        },
      ],
      "FORMACODES": [
        {
          "attributes": {},
          "children": {
            "FORMACODE": [
              {
                "attributes": {},
                "children": {
                  "CODE": [
                    {
                      "attributes": {},
                      "children": {},
                      "value": "42052",
                    },
                  ],
                  "LIBELLE": [
                    {
                      "attributes": {},
                      "children": {},
                      "value": "Maquillage",
                    },
                  ],
                },
              },
            ],
          },
        },
      ],
      "ID_FICHE": [
        {
          "attributes": {
            "yop": "123",
          },
          "children": {},
          "value": "22192",
        },
      ],
      "INTITULE": [
        {
          "attributes": {},
          "children": {},
          "value": "Réaliser un maquillage permanent",
        },
      ],
      "NUMERO_FICHE": [
        {
          "attributes": {},
          "children": {},
          "value": "RS5841",
        },
      ],
    },
  },
  {
    "attributes": {},
    "children": {
      "ID_FICHE": [
        {
          "attributes": {},
          "children": {},
          "value": "22731",
        },
      ],
      "NUMERO_FICHE": [
        {
          "attributes": {},
          "children": {},
          "value": "RS6010",
        },
      ],
    },
  },
]
`;
